### ISSUE TITLE
[feature request] Native ESLint support

### ISSUE THREAD

**Issue by AMorgaut** [SonarSource/SonarJS#871]
**I want to request a feature.**

I'd want a "Native" ESLint support from SonarJS with:

* integrated eslint launch
* integrated Rules support (with list & enable/disable from sonarqube UI)
* bonus: support of eslintrc / eslintignore

### Context

There is plenty or JS linters out there like at the very least:

- [JSLint](http://jslint.com): the original JS Linter from Douglas Crockford
- [JSHint](http://jshint.com/): the first fork due to license & rules disagreements
- [JavaScriptLint](http://www.javascriptlint.com): a more isolated mono-developer project 
- [ESLint](https://eslint.org/): The most active & extensible Linter

To me, Sonar is not only a linter but also a *Continuous Integration* platform with its integrated Lint flavor

At best, I'd love a CI tool supporting any linter and handling for them:

- direct feedback (SonarLint)
- the PASSED/FAILED validation (quality gate), 
- dashboards, 
- and history (not yet very usable in Sonarqube, but that would be another ticket for sonarqube...)

### Technical Context

Usually, Lint is done in 2 steps:

- code parsing (generates an AST: Abstract Syntax Tree)
- looks for rules conformity in this AST

Luckily, the AST format is quite standardized (cf  [ESTree spec](https://github.com/estree/estree)), even if there is few flavors

Main parsers:

- [acorn](https://github.com/ternjs/acorn) from ternjs
- [babylon](https://github.com/babel/babel/tree/master/packages/babylon) from babel, based on acorn
- [espree](https://github.com/eslint/espree) variant used by ESLint, initially based on Esprima (compatible), use acorn
- [esprima](http://esprima.org/) used by istanbul and so many others... but istanbul team moved to the [nyc](https://github.com/istanbuljs/nyc) project with its internal [babel-plugin-istanbul](https://github.com/istanbuljs/babel-plugin-istanbul) so babylon must not be far...
- [cherow](https://github.com/cherow/cherow)

Note also that most of them also support JSX and/or TypeScript syntax

What would be sad is to parse the code for each linter we want to support, but luckily again, it turns out that tools like ESLint (or Babel) provide an API that accept already generated AST

```javascript
const { SourceCode, Lint } = require("eslint");
const linter = new Linter();
const messages = linter.verify(
    new SourceCode(sourceContent, ast), 
    config, // rules, 
    filename
);
```
> https://eslint.org/docs/developer-guide/nodejs-api

### Why ESLint then

While I personally follow, use, and occasionally try to contribute to JSLint, JSHint & ESLint, the later have few advantages that make it a good candidate for SonarLint

1. use a parser generating a ESTree standard compliant output (should be easy to generate a compatible one while SonarJS parse JS / TS files)
2. provides an API that accept already generated AST
3. is extensions oriented and have a large community of contributors making it a good friend of many frameworks
4. It would allow SonarJS to expose many community available rules, fully integrated in the Sonarqube interface (and allow any JS dev to more easily add features)

**Cons**

Of course it may mean that SonarJs would require node.js. But would it be really surprising...

PS:
I am already aware of:
- [sonar-web-frontend-plugin](https://github.com/groupe-sii/sonar-web-frontend-plugin)
- [SonarEsLintPlugin](https://github.com/sleroy/SonarEsLintPlugin)
- and https://github.com/SonarSource/SonarJS/issues/726

**Comment by KFlash** [comment]
You could add [Cherow](https://github.com/cherow/cherow) to that list too. It also use the ESTree standard.

**Comment by mimiz** [comment]
👍

**Comment by penx** [comment]
Our devops team wants our JS code to pass SonarJS rules.

Our JS developers are using eslint, including CLI and IDE addons to highlight issues, which is a fairly standard setup. They would rather not be running 2 linters locally, in part due to Sonar no longer having a supported CLI and encountering issues with some of the Sonar IDE plugins.

To me it would be ideal to have a set of eslint rules that pick up all of the [SonarJS rules](https://www.sonarsource.com/products/codeanalyzers/sonarjs/rules.html), which I am fairly sure is part of what is being requested here.

I think eslint already has rules written that cover most of SonarJS so creating a config for these I would assume to be fairly trivial. There are a few rules that may need porting, such as 'Untrusted content should not be included'.

**Comment by AMorgaut** [comment]
@penx well my initial topic was more to allow the Sonar Js linter to also support ESlint rules and Eslint default integration in many JS projects...

My own motivation is based  on what ESLint does not provide... Meaning the Sonarqube based historic monitoring and manager friendly dashboards and administration... A whole serious CI environment....

If the goal is only to look at the pure lint level... well ESLint has a big advantage on the plugin ecosystem, but on the other side Sonarlint parser raise issues that ESLint doesn't yet proposes.

They are also other differences like between the way Sonar & ESLint estimate code complexity.

The thing is that with ESLint, JSHint, JSlint, or others, most of the time, the best you'll have is a precommit hook that will block commits on error level issues... But all warnings will pass and remain almost untracked to at least all those that refuse to add linter plugins to their IDE (or "pure" code editor). Here Sonarqube comes to help find out about your technical debt with an estimation of how complex the code is for reviewers or developpers that may require some refactoring...

**Comment by penx** [comment]
Ok I've started a package at https://github.com/penx/eslint-config-sonar which perhaps people will find useful if they're in a similar position to what I describe above. I'd encourage @SonarSource to take ownership of this, if it's something they'd like to support going forwards.

**Comment by vilchik-elena** [comment]
Hey here! We have some good news for this thread. Have a look at this ticket https://jira.sonarsource.com/browse/MMF-1231

**Comment by vilchik-elena** [comment]
See https://jira.sonarsource.com/browse/MMF-1409
